File: sample-app-java+push-notification/src/main/java/com/cometchat/sampleapp/java/fcm/voip/CometChatVoIPConnectionService.java
package com.cometchat.sampleapp.java.fcm.voip;

import android.os.Bundle;
import android.os.Handler;
import android.telecom.Connection;
import android.telecom.ConnectionRequest;
import android.telecom.ConnectionService;
import android.telecom.DisconnectCause;
import android.telecom.PhoneAccountHandle;
import android.telecom.TelecomManager;

public class CometChatVoIPConnectionService extends ConnectionService {
    private static final int CALL_TIMEOUT = 40000; // 40 seconds
    private Handler timeoutHandler;
    private Runnable timeoutRunnable;

    @Override
    public Connection onCreateIncomingConnection(PhoneAccountHandle connectionManagerPhoneAccount, ConnectionRequest request) {
        CometChatVoIPUtils.isCallOngoing = true;
        Bundle extras = request.getExtras();
        String callerName = extras.getString("callerName", "Unknown Caller");
        String callerAvatar = extras.getString("callerAvatar", "");
        CometChatVoIPConnection connection = new CometChatVoIPConnection(this);
        connection.setExtras(extras);
        connection.setCallerDisplayName(callerName, TelecomManager.PRESENTATION_ALLOWED);
        connection.setAddress(request.getAddress(), TelecomManager.PRESENTATION_ALLOWED);
        connection.setRinging();

        // Initialize handler and timeout runnable
        timeoutHandler = new Handler();
        timeoutRunnable = () -> {
            if (connection.getState() == Connection.STATE_RINGING) {
                connection.setDisconnected(new DisconnectCause(DisconnectCause.REJECTED));
                connection.destroy();
            }
        };
        timeoutHandler.postDelayed(timeoutRunnable, CALL_TIMEOUT);

        return connection;
    }

    @Override
    public Connection onCreateOutgoingConnection(PhoneAccountHandle connectionManagerPhoneAccount, ConnectionRequest request) {
        return new CometChatVoIPConnection(this);
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        if (timeoutHandler != null) {
            timeoutHandler.removeCallbacks(timeoutRunnable);
        }
        CometChatVoIPUtils.isCallOngoing = false;
    }


}

File: sample-app-kotlin+push-notification/src/main/java/com/cometchat/sampleapp/kotlin/fcm/voip/CometChatVoIPConnectionService.kt
package com.cometchat.sampleapp.kotlin.fcm.voip

class CometChatVoIPConnectionService : android.telecom.ConnectionService() {
    private var timeoutHandler: android.os.Handler? = null
    private var timeoutRunnable: Runnable? = null

    override fun onCreateIncomingConnection(
        connectionManagerPhoneAccount: android.telecom.PhoneAccountHandle, request: android.telecom.ConnectionRequest
    ): android.telecom.Connection {
        CometChatVoIPUtils.isCallOngoing = true
        val extras = request.extras
        val callerName = extras.getString("callerName", "Unknown Caller")
        val callerAvatar = extras.getString("callerAvatar", "")
        val connection = CometChatVoIPConnection(this)
        connection.extras = extras
        connection.setCallerDisplayName(callerName, android.telecom.TelecomManager.PRESENTATION_ALLOWED)
        connection.setAddress(request.address, android.telecom.TelecomManager.PRESENTATION_ALLOWED)
        connection.setRinging() // Initialize handler and timeout runnable
        timeoutHandler = android.os.Handler()
        timeoutRunnable = Runnable {
            if (connection.state == android.telecom.Connection.STATE_RINGING) {
                connection.setDisconnected(android.telecom.DisconnectCause(android.telecom.DisconnectCause.REJECTED))
                connection.destroy()
            }
        }
        timeoutHandler!!.postDelayed(timeoutRunnable!!, CALL_TIMEOUT.toLong())

        return connection
    }

    override fun onCreateOutgoingConnection(
        connectionManagerPhoneAccount: android.telecom.PhoneAccountHandle, request: android.telecom.ConnectionRequest
    ): android.telecom.Connection {
        return CometChatVoIPConnection(this)
    }

    override fun onDestroy() {
        super.onDestroy()
        if (timeoutHandler != null) {
            timeoutHandler!!.removeCallbacks(timeoutRunnable!!)
        }
        CometChatVoIPUtils.isCallOngoing = false
    }

    companion object {
        private const val CALL_TIMEOUT = 40000 // 40 seconds
    }
}